BlueprintUI/Sources/Internal/CacheTree.swift

```diff
@@ -0,0 +1,52 @@
+import CoreGraphics
+import os.log
+
+/// A size cache that also holds subcaches.
+protocol CacheTree: AnyObject {
+
+    /// The name of this cache
+    var name: String { get }
+
+    /// A reference to use for logging
+    var signpostRef: AnyObject { get }
+
+    /// The sizes that are contained in this cache, keyed by size constraint.
+    subscript(constraint: SizeConstraint) -> CGSize? { get set }
+
+    /// Gets a subcache identified by the given key, or creates a new one.
+    func subcache(key: SubcacheKey, name: @autoclosure () -> String) -> CacheTree
+}
+
+struct SubcacheKey: RawRepresentable, Hashable {
+    /// A key indicating that this will be the only subcache
+    static let singleton = SubcacheKey(rawValue: -1)
```

Comment:
Should we do this, or would something like this be clearer?

```
enum SubcacheKey : Hashable {
   case singleton
   case indexed(Int)
}
```

Reply:
> Should we do this, or would something like this be clearer?
> 
> ```
> enum SubcacheKey : Hashable {
>    case singleton
>    case indexed(Int)
> }
> ```

Maybe? I was torn on how much to optimize this one. The singleton case is only used to improve the element name emitted to signpost, so we could also get rid of this type entirely and just pass in a hint along with the name.